src/NServiceBus.Core/Transports/FileBased/MessagePump.cs

```diff
@@ -0,0 +1,195 @@
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus.Extensibility;
+    using NServiceBus.Logging;
+
+    class MessagePump : IPushMessages
+    {
+        static ILog Logger = LogManager.GetLogger<MessagePump>();
+
+        public void Init(Func<PushContext, Task> pipe, PushSettings settings)
+        {
+            pipeline = pipe;
+            path = Path.Combine("c:\\bus", settings.InputQueue);
+            purgeOnStartup = settings.PurgeOnStartup;
+        }
+
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            runningReceiveTasks = new ConcurrentDictionary<Task, Task>();
+            concurrencyLimiter = new SemaphoreSlim(limitations.MaxConcurrency);
+            cancellationTokenSource = new CancellationTokenSource();
+
+            cancellationToken = cancellationTokenSource.Token;
+
+            if (purgeOnStartup)
+            {
+                Directory.Delete(path, true);
+                Directory.CreateDirectory(path);
+            }
+
+            messagePumpTask = Task.Factory.StartNew(() => ProcessMessages(), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
+        }
+
+        public async Task Stop()
+        {
+            cancellationTokenSource.Cancel();
+
+            // ReSharper disable once MethodSupportsCancellation
+            var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
+            var allTasks = runningReceiveTasks.Values.Concat(new[]
+            {
+                messagePumpTask
+            });
+            var finishedTask = await Task.WhenAny(Task.WhenAll(allTasks), timeoutTask).ConfigureAwait(false);
+
+            if (finishedTask.Equals(timeoutTask))
+            {
+                Logger.Error("The message pump failed to stop with in the time allowed(30s)");
+            }
+
+            concurrencyLimiter.Dispose();
+            runningReceiveTasks.Clear();
+        }
+
+        [DebuggerNonUserCode]
+        async Task ProcessMessages()
+        {
+            try
+            {
+                await InnerProcessMessages().ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // For graceful shutdown purposes
+            }
+            catch (Exception ex)
+            {
+                Logger.Error("File Message pump failed", ex);
+                //await peekCircuitBreaker.Failure(ex).ConfigureAwait(false);
+            }
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                await ProcessMessages().ConfigureAwait(false);
+            }
+        }
+
+        async Task InnerProcessMessages()
+        {
+            while (!cancellationTokenSource.IsCancellationRequested)
+            {
+                var filesFound = false;
+
+                foreach (var filePath in Directory.EnumerateFiles(path, "*.*"))
```

Comment:
Maybe it would be good to add some ordering to the way we pick up files? (we could do 'goes to the end of queue' for rollbacked messages